spatial/curve/hilbert.go

```diff
@@ -0,0 +1,241 @@
+// Copyright ©2024 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package curve
+
+// Hilbert2D is a 2-dimensional Hilbert curve.
+type Hilbert2D struct{ Order int }
+
+// Dims returns the spatial dimensions of the curve, which is {2ᵏ, 2ᵏ}, where k
+// is the order.
+func (h Hilbert2D) Dims() []int { return []int{1 << h.Order, 1 << h.Order} }
+
+// Len returns the length of the curve, which is 2ⁿᵏ, where n is the dimension
+// (2) and k is the order.
+//
+// Len will overflow on a 32-bit architecture if the order is ≥ 16.
```

Comment:
I updated these comments to "on architectures where int is 32-bits" instead of "32-bit architecture" and added a comment for "where int is 64-bits".

Reply:
I don't think that makes a difference; we can still see an overflow with b4-bit ints.

Reply:
> I do think it would be better to include an equivalent warning on the constructor

What constructor? I could add `func NewHilbert(order, dimension int) (???, error)` but that would require something like `type Curve interface`. Or I could add `func NewHilbert2D(order int) (Hilbert2D, error)` etc but the implementation would be trivial (not worth a function IMO).

Reply:
I was thinking func `NewHilbert2D(order int) (Hilbert2D, error)` and friends. The function is trivial, but makes the overflow checks documented in code.

Reply:
Done